amaranth/hdl/_dsl.py

```diff
@@ -288,8 +288,11 @@ def __init__(self):
         self._domains      = {}
         self._generated    = {}
         self._src_loc      = tracer.get_src_loc()
+        self._frozen       = False
 
     def _check_context(self, construct, context):
+        if self._frozen:
+            raise AlreadyElaborated(f"Cannot use control structures on a module that has already been elaborated")
```

Comment:
We definitely don't need unique diagnostics for all of these cases, as it's the same underlying issue and it is unimportant how the module is modified. (Otherwise you might conclude that there are some other modifications that are permitted, whereas none are.)

```suggestion
            raise AlreadyElaborated(f"Cannot modify a module that has already been elaborated")
```